File: backend/app/utils/national_id.py
import re


def validate_iranian_national_id(national_id: str) -> bool:
    """Validate Iranian national ID (10 digits + standard checksum)."""
    if not re.fullmatch(r"\d{10}", national_id):
        return False

    if len(set(national_id)) == 1:
        return False

    check_digit = int(national_id[9])
    weighted_sum = sum(int(national_id[i]) * (10 - i) for i in range(9)) % 11

    if weighted_sum < 2:
        return check_digit == weighted_sum
    return check_digit == 11 - weighted_sum

File: frontend/lib/national-id.ts
/** Iranian national ID validation (10 digits + standard checksum). */
export function validateIranianNationalId(nationalId: string): boolean {
  if (!/^\d{10}$/.test(nationalId)) return false;
  if (new Set(nationalId).size === 1) return false;

  const checkDigit = Number(nationalId[9]);
  const weightedSum =
    nationalId
      .slice(0, 9)
      .split("")
      .reduce((sum, digit, index) => sum + Number(digit) * (10 - index), 0) % 11;

  if (weightedSum < 2) return checkDigit === weightedSum;
  return checkDigit === 11 - weightedSum;
}

export const AUTH_ERROR_MESSAGES: Record<string, string> = {
  USER_EXISTS: "این کاربر قبلاً ثبت‌نام کرده است",
  INVALID_NATIONAL_ID: "کد ملی وارد شده معتبر نیست",
};

export function mapAuthError(detail: string | undefined, fallback: string): string {
  if (!detail) return fallback;
  return AUTH_ERROR_MESSAGES[detail] ?? detail;
}
